oracle/src/lib.rs

```diff
@@ -552,11 +573,43 @@ mod tests {
 
         contract.request(payment.clone(), spec_id.clone(), callback_address.clone(), callback_method.clone(), nonce.clone(), data_version.clone(), data.clone());
         context.prepaid_gas = 10u64.pow(18);
-        contract.store_request( alice(), payment.clone(), spec_id.clone(), callback_address.clone(), callback_method.clone(), nonce.clone(), data_version.clone(), data.clone());
+        contract.store_request(alice(), payment.clone(), spec_id.clone(), callback_address.clone(), callback_method.clone(), nonce.clone(), data_version.clone(), data.clone());
         testing_env!(context.clone());
 
         contract.request(payment.clone(), spec_id.clone(), callback_address.clone(), callback_method.clone(), nonce.clone(), data_version.clone(), data.clone());
-        contract.store_request( alice(), payment, spec_id, callback_address, callback_method, nonce, data_version, data);
+        contract.store_request(alice(), payment, spec_id, callback_address, callback_method, nonce, data_version, data);
+    }
+
+    #[test]
+    #[should_panic(
+        expected = "Invalid, already used nonce: 7"
+    )]
+    fn make_invalid_nonce_request() {
+        let mut context = get_context(alice(), 0);
+        context.attached_deposit = TRANSFER_FROM_NEAR_COST;
+        testing_env!(context.clone());
+        let mut contract = Oracle::new(link(), alice());
+        let payment: U128 = 6_u128.into();
+        let spec_id = encode("unique spec id".to_string());
+        let callback_address = "callback.testnet".to_string();
+        let callback_method = "test_callback".to_string();
+        let data_version: U128 = 131_u128.into();
+        let data = encode("BAT".to_string());
+
+        contract.request(payment.clone(), spec_id.clone(), callback_address.clone(), callback_method.clone(), 8_u128.into(), data_version.clone(), data.clone());
+        context.prepaid_gas = 10u64.pow(18);
+        contract.store_request(alice(), payment.clone(), spec_id.clone(), callback_address.clone(), callback_method.clone(), 8_u128.into(), data_version.clone(), data.clone());
+        testing_env!(context.clone());
+
+        let default: U128 = 0_u128.into();
+        let current_nonce: u128 = contract.get_nonce(alice()).unwrap_or(default).into();
+        assert_eq!(current_nonce, 8_u128.into());
+
+        let current_mapped_nonce: U128 = *contract.get_nonces().get(&alice()).clone().unwrap_or(&default);
```

Comment:
I'm unclear why this wants to use `get_nonces` and turn this into a HashMap
I would expect this to be:
`contract.nonces.get(&alice…`

Reply:
This is a test. We test the publicly exposed function, not the TreeMap implementation.

Reply:
Okay thanks, I thought this might have been the other way around. Knowing that this is needed for the initiator start clears this up.